lib/functoria.ml

```diff
@@ -21,39 +21,147 @@ open Functoria_misc
+
+  let m_option f a b = match a, b with
+    | None, None -> None
+    | Some a, None -> Some a
+    | None, Some b -> Some b
+    | Some a, Some b -> Some (f a b)
+
+  let merge opam a b =
+    let ocamlfind = String.Set.union a.ocamlfind b.ocamlfind
+    and min =
+      m_option
+        (fun a b -> match compare_version a b with 0 -> a | 1 -> a | _ -> b)
+        a.min b.min
+    and max =
+      m_option
+        (fun a b -> match compare_version a b with 0 -> a | 1 -> b | _ -> a)
+        a.max b.max
+    in
+    match min, max with
+    | Some a, Some b when compare_version a b >= 0 -> invalid_arg "version constraint must be that min is smaller than max"
+    | _ -> Some { opam ; build = a.build ; ocamlfind ; min ; max }
+
+  let package ?(build = false) ?sublibs ?ocamlfind ?min ?max opam =
+    let ocamlfind = match sublibs, ocamlfind with
+      | None, None -> [opam]
+      | Some xs, None -> opam :: List.map (fun x -> opam ^ "." ^ x) xs
+      | None, Some a -> a
+      | Some _, Some _ -> invalid_arg "only either ~sub or ~ocamlfind may be specified"
+    in
+    let ocamlfind = String.Set.of_list ocamlfind in
+    match min, max with
+    | Some min, Some max when compare_version min max >= 0 -> invalid_arg "min must be < max"
+    | _ -> { opam ; build ; ocamlfind ; min ; max }
+end
+
+let package = Package.package
 
+module Info = struct
   type t = {
     name: string;
     root: string;
     keys: Key.Set.t;
     context: Key.context;
-    libraries: String.Set.t;
-    packages: String.Set.t;
+    packages: package String.Map.t;
   }
 
   let name t = t.name
   let root t = t.root
-  let libraries t = String.Set.elements t.libraries
-  let packages t = String.Set.elements t.packages
+  let libraries t =
+    String.Set.elements
+      (List.fold_left String.Set.union String.Set.empty
+         (List.map (fun x -> x.ocamlfind)
+            (List.filter (fun x -> x.build = false)
+               (List.map snd (String.Map.bindings t.packages)))))
+  let package_names t =
+    List.map fst
+      (List.filter (fun (_, p) -> p.build = false)
+         (String.Map.bindings t.packages))
+  let packages t = List.map snd (String.Map.bindings t.packages)
   let keys t = Key.Set.elements t.keys
   let context t = t.context
 
-  let create ?(packages=[]) ?(libraries=[]) ?(keys=[]) ~context ~name ~root =
-    let libraries = String.Set.of_list libraries in
-    let packages = String.Set.of_list packages in
+  let create ~packages ~keys ~context ~name ~root =
     let keys = Key.Set.of_list keys in
-    { name; root; keys; libraries; packages; context }
-
-  let pp verbose ppf { name ; root ; keys ; context ; libraries ; packages } =
+    let packages = List.fold_left (fun m p ->
+        let n = p.opam in
+        match String.Map.find p.opam m with
+        | None -> String.Map.add n p m
+        | Some p' -> match Package.merge p.opam p p' with
+          | Some p -> String.Map.add n p m
+          | None -> invalid_arg "bad constraints")
```

Comment:
We should provide a better error message (and in particular, which opam package triggers the conflict).